clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+            this.clientKerberosName = new KerberosName(clientPrincipal.getName());
+            // assume that server and client are in the same realm (by default; unless the config
+            // "kafka.server.realm" is set).
+            if (configs.containsKey(SaslConfigs.SASL_KAFKA_SERVER_REALM))
+                this.serverRealm = (String) configs.get(SaslConfigs.SASL_KAFKA_SERVER_REALM);
+            else
+                this.serverRealm = clientKerberosName.realm();
+            this.serviceKerberosName = new KerberosName(servicePrincipal + "@" + serverRealm);
+            this.clientPrincipalName = clientKerberosName.toString();
+            this.saslClient = createSaslClient();
+        } catch (Exception e) {
+            throw new KafkaException("Failed to configure SaslClientAuthenticator", e);
+        }
+    }
+
+    private SaslClient createSaslClient() {
+        try {
+            saslClient = Subject.doAs(subject, new PrivilegedExceptionAction<SaslClient>() {
+                    public SaslClient run() throws SaslException {
+                        LOG.debug("Client will use GSSAPI as SASL mechanism.");
+                        String[] mechs = {"GSSAPI"};
+                        LOG.debug("creating sasl client: client=" + clientPrincipalName + ";service=" + servicePrincipal + ";serviceHostname=" + host);
+                        SaslClient saslClient = Sasl.createSaslClient(mechs, clientPrincipalName, servicePrincipal, host, null, new ClientCallbackHandler(null));
+                        return saslClient;
+                    }
+                });
+            return saslClient;
+        } catch (Exception e) {
+            LOG.error("Exception while trying to create SASL client", e);
+            throw new KafkaException("Failed to create SASL client", e);
+        }
+    }
+
+
+    public void authenticate() throws IOException {
+        if (netOutBuffer != null && !flushNetOutBuffer()) {
+            transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+            return;
+        }
+        byte[] serverToken = new byte[0];
+        switch (saslState) {
+            case INITIAL:
+                sendSaslToken(serverToken);
+                saslState = SaslState.INTERMEDIATE;
+                break;
+            case INTERMEDIATE:
+                if (netInBuffer == null) netInBuffer = new NetworkReceive(node);
+                long readLen = netInBuffer.readFrom(transportLayer);
+                if (netInBuffer.complete()) {
+                    netInBuffer.payload().rewind();
+                    serverToken = new byte[netInBuffer.payload().remaining()];
+                    netInBuffer.payload().get(serverToken, 0, serverToken.length);
+                    netInBuffer = null; // reset the networkReceive as we read all the data.
+                    sendSaslToken(serverToken);
+                }
+                if (saslClient.isComplete())
+                    saslState = SaslState.COMPLETE;
```

Comment:
It is probably not enough to just turn off OP_WRITE at SASL completion time. After completely sending a challenge token, the client needs to turn off OP_WRITE. Otherwise, while waiting to receive the next token from the server, the client will be busy checking in the selector.